pkg/plugins/predicate/predicate.go

```diff
@@ -1,21 +1,42 @@
-package scheduling
+package predicate
 
 import (
+	"context"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 
 	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
 	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	"open-cluster-management.io/placement/pkg/plugins"
 )
 
+var _ plugins.Filter = &Predicate{}
+
+type Predicate struct {
+	predicates []clusterapiv1alpha1.ClusterPredicate
+}
+
 type predicateSelector struct {
 	labelSelector labels.Selector
 	claimSelector labels.Selector
 }
 
-// matchWithClusterPredicates returns a slice of clusters. Each of them must match at least one of the predicates.
-func matchWithClusterPredicates(predicates []clusterapiv1alpha1.ClusterPredicate, clusters []*clusterapiv1.ManagedCluster) ([]*clusterapiv1.ManagedCluster, error) {
-	if len(predicates) == 0 {
+func New(handle plugins.Handle) *Predicate {
+	return &Predicate{}
+}
+
+func (p *Predicate) Name() string {
+	return "predicate"
+}
+
+func (p *Predicate) Set(pl *clusterapiv1alpha1.Placement) {
+	p.predicates = pl.Spec.Predicates
```

Comment:
this is looking weird to me.  It almost seems like `.Set` would be called and later `filter` would be called and the order between the calls matters.